From: Chad Dupuis <cdupuis@redhat.com>
Date: Wed, 1 Sep 2010 19:27:25 -0400
Subject: [scsi] qla2xxx: clear post-uncorrectable non-fatal errors
Message-id: <20100901192716.2304.82946.sendpatchset@localhost.localdomain>
Patchwork-id: 27999
O-Subject: [RHEL 5.6 PATCH] qla2xxx: Clear error status after uncorrectable
	non-fatal errors.
Bugzilla: 572258
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

qla2xxx: Clear error status after uncorrectable non-fatal errors.

Bugzilla
--------

Bug 572258 (https://bugzilla.redhat.com/show_bug.cgi?id=572258)

Upstream Status
---------------

scsi-misc commit id 3e46f031322bdd8d1f11e17fb7cf00c38b08dd55

Brew Build
----------

2726652

Testing
-------

This patch was test built against the 2.6.18-215.el5 kernel.  This patch was
functionally tested at QLogic.

Description
-----------

>From 8f7dd0ea8d0c930bd6d5b62a97bde8e423e12882 Mon Sep 17 00:00:00 2001
From: Lalit Chandivade <lalit.chandivade@qlogic.com>
Date: Wed, 17 Mar 2010 16:54:56 +0530
Subject: [PATCH] qla2xxx: Clear error status after uncorrectable non-fatal errors.

Currently error status is cleared only after the uncorrectable fatal errors
in the qla2xxx_pci_slot_reset. This fix is added to clear the error status in
qla2xxx_pci_resume. This way for both fatal and non-fatal errors the error
status gets cleared properly.

diff --git a/drivers/scsi/qla2xxx/qla_os.c b/drivers/scsi/qla2xxx/qla_os.c
index 0d1438e..bcf97cd 100644
--- a/drivers/scsi/qla2xxx/qla_os.c
+++ b/drivers/scsi/qla2xxx/qla_os.c
@@ -3901,7 +3901,6 @@ qla2xxx_pci_slot_reset(struct pci_dev *pdev)
 		ret =  PCI_ERS_RESULT_RECOVERED;
 	clear_bit(ABORT_ISP_ACTIVE, &ha->dpc_flags);
 
-	pci_cleanup_aer_uncorrect_error_status(pdev);
 
 exit_slot_reset:
 	qla_printk(KERN_WARNING, ha, "slot_reset-return:ret=%x\n",ret);
@@ -3924,6 +3923,8 @@ qla2xxx_pci_resume(struct pci_dev *pdev)
 		    "from slot/link_reset");
 	}
 
+	pci_cleanup_aer_uncorrect_error_status(pdev);
+
 	ha->flags.eeh_busy = 0;
 }
 
